common/models/templates/test/test-trie-builder.js

```diff
@@ -0,0 +1,257 @@
+    assert.equal(defaultRoot.weight, 5);
+
+    // Adds a second child; that child has more than one letter not-in-common with the first.
+    trieConstruction.addUnsorted(defaultRoot, {
+      content: 'call',
+      weight: 8,
+      key: 'call'
+    }, 2);
+
+    assert.equal(defaultRoot.type, 'internal');
+    assert.equal(defaultRoot.weight, 8);
+    assert.sameMembers(defaultRoot.values, ['f', 'l']);
+    assert.equal(defaultRoot.children['f'].weight, 5);
+    assert.equal(defaultRoot.children['l'].weight, 8);
+  });
+
+  it('addUnsorted - lower frequency prefix', () => {
+    // Will correspond to 'thin'.
+    const root = trieConstruction.createRootNode();
+    trieConstruction.addUnsorted(root, {
+      key: 'think',
+      content: 'think',
+      weight: 100
+    }, 4);
+    trieConstruction.addUnsorted(root, {
+      key: 'thing',
+      content: 'thing',
+      weight: 40
+    }, 4);
+
+    assert.equal(root.type, 'internal');
+    assert.sameMembers(root.values, ['k', 'g']);
+
+    // Interesting noted behavior:  if just 'think' into 'thin', and nothing else...
+    // it remains a leaf!  That doesn't seem... entirely proper.  With 'thing', it
+    // will at least be an 'internal' node already.
+    trieConstruction.addUnsorted(root, {
+      key: 'thin',
+      content: 'thin',
+      weight: 20
+    }, 4);
+
+    assert.sameMembers(root.values, ['k', 'g', SENTINEL_CODE_UNIT]);
+    assert.isOk(root.children[SENTINEL_CODE_UNIT]);
+    assert.equal(root.children[SENTINEL_CODE_UNIT].weight, 20);
+  });
+
+  describe('TrieBuilder', () => {
+    it('standard Trie construction', () => {
+      const builder = new TrieBuilder((text) => text);
+      builder.addEntry('caffeine', 2);
+      builder.addEntry('cafe', 5);
+      builder.addEntry('calm', 3);
+      builder.addEntry('calf', 4);
+      builder.addEntry('call', 6); // total: 20
+      builder.addEntry('can', 10); // total: 30
+      builder.addEntry('and', 20); // total: 50
+
+      assert.equal(builder.getTotalWeight(), 50);
+
+      const root = builder.getRoot();
+      const aNode = root.children['c'].children['a'];
+
+      // As the nodes were not added in sorted order, they should not currently be ordered.
+      assert.sameMembers(aNode.values, ['n', 'l', 'f']);
+      assert.notSameOrderedMembers(aNode.values, ['n', 'l', 'f']);
+
+      const lNode = aNode.children['l'];
+      assert.sameMembers(lNode.values, ['l', 'f', 'm']);
+      assert.notSameOrderedMembers(lNode.values, ['l', 'f', 'm']);
+
+      builder.sort();
+
+      assert.sameOrderedMembers(aNode.values, ['n', 'l', 'f']);
+      assert.sameOrderedMembers(lNode.values, ['l', 'f', 'm']);
+    });
+
+    // In case of high startup-time for user-dictionary processing, this could allow use
+    // of a 'partially' processed user wordlist.  (We'd prioritize predicting when the
+    // user is interacting with text, then resume processing once 'idle'.)
+    it('interspersed construction + lookup', () => {
```

Comment:
Admittedly, this was written entirely before recent discussions some of the team and I have had offline.  It sounds like we probably won't leverage this, but it _is_ ready; figured the least I can do it let it show up in review (and `git` history), even should we axe it.